design.md

```diff
@@ -0,0 +1,273 @@
+# Kernel Design
+
+## Functional Requirements
+
+### gRPC Server
+
+The kernel runs a gRPC server that supports the following operations.
+
+#### Installation and Updating Operations
+
+- Add a systemd service that starts the kernel on boot.
+- Enable/disable that service.
+- Transfer the kernel Jar and return the file path.
+  - This call will be used when the kernel is remote. If the kernel is on the same machine as the client, then the client will handle putting the new Jar into some file and getting the file path.
+- Update the kernel using a file path.
+  - This call will be used whether the kernel is on the same machine as the client or whether it is remote.
+
+#### Script Operations
+
+- Run a script from a file specifier and a list of dev'd libraries
+
+#### GitFS Operations
+
+- Clone or pull a repo.
+- List files in a repo.
+- Read the contents of a file in a repo.
+- Clear the cache.
+
+#### Hardware Operations
+
+- Load a robot from its config file (given as a file specifier) and a connection method.
+
+#### Plugin Operations
+
+- List plugins with a matching group, name, and version.
+  - The group and name fields may be matched using a regex.
+  - The version field may be matched using a version range.
+- Download a plugin given its triple.
+- Clear the plugin cache.
+
+#### Client Interaction
+
+- The client and kernel share a bidirectional stream.
+  - The kernel can create tasks. Task schema:
+    - Unique ID
+    - Description
+    - Progress
+  - The kernel can update tasks. Update schema:
+    - Task ID
+    - Progress
+  - The kernel can stop tasks. Stop schema:
+    - Task ID
+  - The kernel can ask the IDE for confirmation. Confirmation schema:
+    - Description
+    - Confirmation response: Allowed/Denied
+  - A progress of `NaN` must show an indeterminate progress bar. Otherwise, progress is a percentage stored as a floating point number in the range `[0, 100]`.
+  - The kernel can ask the IDE for credentials. The IDE may respond with credentials or may deny the request.
+
+#### Timeout
+
+The client must periodically call a keepalive function in the kernel. If the kernel has not received a keepalive call for some time period, then the kernel must interrupt any running scripts, disconnect any connected hardware, and exit. If the kernel is running headlessly without a client, then no keepalives are required to keep the kernel alive.
+
+- When started headlessly, the kernel is configured to not have a timeout.
+- When started with a head, the kernel is configured to have a timeout.
+
+#### Display Operations
+
+- Server streaming RPC.
+- Argument provided by client indicates which series to respond to
+  - ! RFC: how do we identify the series?
```

Comment:
Data series (sometimes called "data sources") have a unique ID, a name, and a type.